protocols/gossipsub/src/handler.rs

```diff
@@ -564,17 +516,58 @@ impl ConnectionHandler for Handler {
             Self::OutboundOpenInfo,
         >,
     ) {
+        if event.is_outbound() {
+            self.outbound_substream_establishing = false;
+        }
+
+        if event.is_inbound() && self.inbound_substreams_created == MAX_SUBSTREAM_CREATION {
+            // Too many inbound substreams have been created, disable the handler.
+            self.keep_alive = KeepAlive::No;
+            log::info!("The maximum number of inbound substreams created has been exceeded.");
+            return;
+        }
+
+        if event.is_outbound() && self.outbound_substreams_created == MAX_SUBSTREAM_CREATION {
+            // Too many outbound substreams have been created, disable the handler.
+            self.keep_alive = KeepAlive::No;
+            log::info!("The maximum number of outbound substreams created has been exceeded.");
+            return;
+        }
+
         match event {
             ConnectionEvent::FullyNegotiatedInbound(fully_negotiated_inbound) => {
                 self.on_fully_negotiated_inbound(fully_negotiated_inbound)
             }
             ConnectionEvent::FullyNegotiatedOutbound(fully_negotiated_outbound) => {
                 self.on_fully_negotiated_outbound(fully_negotiated_outbound)
             }
-            ConnectionEvent::DialUpgradeError(DialUpgradeError { error: e, .. }) => {
-                self.outbound_substream_establishing = false;
-                warn!("Dial upgrade error {:?}", e);
-                self.upgrade_errors.push_back(e);
+            ConnectionEvent::DialUpgradeError(DialUpgradeError {
+                error: ConnectionHandlerUpgrErr::Timeout | ConnectionHandlerUpgrErr::Timer,
+                ..
+            }) => {
+                log::debug!("Dial upgrade error: Protocol negotiation timeout.");
+            }
+            ConnectionEvent::DialUpgradeError(DialUpgradeError {
+                error: ConnectionHandlerUpgrErr::Upgrade(UpgradeError::Apply(e)),
+                ..
+            }) => void::unreachable(e),
+            ConnectionEvent::DialUpgradeError(DialUpgradeError {
+                error:
+                    ConnectionHandlerUpgrErr::Upgrade(UpgradeError::Select(NegotiationError::Failed)),
+                ..
+            }) => {
+                // The protocol is not supported
+                self.protocol_unsupported = true;
+                log::debug!("The remote peer does not support gossipsub on this connection");
+            }
+            ConnectionEvent::DialUpgradeError(DialUpgradeError {
+                error:
+                    ConnectionHandlerUpgrErr::Upgrade(UpgradeError::Select(
+                        NegotiationError::ProtocolError(e),
+                    )),
+                ..
+            }) => {
+                log::debug!("Protocol negotiation failed: {e}")
```

Comment:
Similarly here. ARe we going to constantly try and re-negotiate? I think we should either drop the connection of set KeepAlive to No.

Reply:
Up to `MAX_SUBSTREAM_CREATION` times only, then the handler will be set to `KeepAlive::No`.